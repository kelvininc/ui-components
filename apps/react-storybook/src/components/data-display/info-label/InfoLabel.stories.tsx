
import type { Meta, StoryFn, StoryObj } from '@storybook/react';
import { KvInfoLabel, KvTagLetter } from '@kelvininc/react-ui-components';
import { ComponentProps } from 'react';

const meta = {
	title: 'Data Display/Info Label',
	component: KvInfoLabel,
	argTypes: {
		labelTitle: { control: { type: 'text' } },
		description: { control: { type: 'text' } },
		descriptionHeight: { control: { type: 'number' } },
		descriptionCollapsedText: { control: { type: 'text' } },
		descriptionOpenedText: { control: { type: 'text' } },
		copyValue: { control: { type: 'text' } }
	}
} satisfies Meta<typeof KvInfoLabel>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
	args: {
		labelTitle: 'TYPE',
		description: 'Data Model'
	}
};


export const ReadMoreLess: Story = {
	args: {
		labelTitle: 'DESCRIPTION',
		description:
			"Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
		descriptionHeight: 34,
		descriptionCollapsedText: 'Read more',
		descriptionOpenedText: 'Read less'
	}
};


export const CopyValues: Story = {
	args: {
		labelTitle: 'TYPE',
		description: 'data-model',
		copyValue: 'data-model'
	}
};


const InfoLabelTemplateWithComponent: StoryFn<ComponentProps<typeof KvInfoLabel>> = args => (
	<KvInfoLabel {...args}>
		<KvTagLetter tagLetter="C" label="Change The Rapper" />
	</KvInfoLabel>
);

export const WithComponent: Story = {
	args: {
		labelTitle: 'DESCRIPTION'
	},
	render: InfoLabelTemplateWithComponent
};
