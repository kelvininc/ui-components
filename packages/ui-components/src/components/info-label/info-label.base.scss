@import '../../assets/styles/globals';

/**
	 * @prop --text-color-title: Info label's title color.
	 * @prop --text-color-description: Info label's description color.
	 * @prop --expanded-buttom-color: Expandded buttom's color.
	 * @prop --description-fade-height: Description fade height.
	 * @prop --description-fade-shadow: Description fade shadow.
*/

:host {
	--text-color-title: #{kv-color('neutral-4')};
	--text-color-description: #{kv-color('neutral-2')};
	--expanded-button-color: #{kv-color('neutral-0')};
	--description-fade-height: 30px;
	--description-fade-shadow: linear-gradient(to bottom, rgba(32, 32, 32, 0) 0%, rgba(32, 32, 32, 1) 100%);
}

.info-label {
	display: flex;
	flex-direction: column;

	.title {
		@include kv-font-label-small-uppercase-regular();
		color: var(--text-color-title);
		white-space: nowrap;
		margin-bottom: $spacing-2x;

		&.no-description {
			margin-bottom: 0;
		}
	}

	.description-wrapper {
		position: relative;
		overflow-y: hidden;
		transition: height 0.3s ease-in-out;
		height: auto;

		.text {
			@include kv-font-span-regular();
			color: var(--text-color-description);
		}
	}

	.expand-description-button {
		@include kv-font-label-small-regular();
		display: flex;
		align-items: center;
		cursor: pointer;
		margin-top: $spacing-3x;
		color: var(--expanded-button-color);

		kv-icon {
			--icon-color: #{var(--expanded-button-color)};
			margin-left: $spacing;
			transition: transform 0.3s ease-in-out;
		}

		&.expanded kv-icon {
			--icon-rotation: 180deg;
		}
	}
}

.description {
	display: flex;
	align-items: center;

	.copy-icon {
		opacity: 0;
	}

	&:hover .copy-icon {
		opacity: 1;
		transition: all 200ms ease-in-out;
	}
}

.copy-icon {
	margin-left: $spacing-2x;
	cursor: pointer;
	user-select: none;
}

.description-fade {
	position: absolute;
	bottom: -2px;
	left: 0;
	right: 0;
	height: var(--description-fade-height);
	background: var(--description-fade-shadow);
	pointer-events: none;
	transition: background 0.3s ease-in-out;

	&--expanded {
		background: transparent;
	}
}
