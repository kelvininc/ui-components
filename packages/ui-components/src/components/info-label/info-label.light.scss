@import '../../assets/styles/globals';
@import './info-label.base';


:host {
	--text-color-title: #{kv-color('text', 'step-3')};
	--text-color-description: #{kv-color('text', 'step-5')};
	--expanded-button-color: #{kv-color('primary')};
}
