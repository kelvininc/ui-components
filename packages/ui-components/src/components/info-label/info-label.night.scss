@import '../../assets/styles/globals';
@import './info-label.base';

:host {
	--text-color-title: #{kv-color('neutral-4')};
	--text-color-description: #{kv-color('neutral-2')};
	--expanded-button-color: #{kv-color('neutral-0')};
	--description-fade-shadow: linear-gradient(to bottom, rgba(32, 32, 32, 0) 0%, rgba(32, 32, 32, 1) 100%);
}
