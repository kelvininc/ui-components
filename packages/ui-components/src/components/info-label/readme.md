# *<kv-info-label>*



<!-- Auto Generated Below -->


## Usage

### React

```tsx
import React from 'react';

import { KvInfoLabel, KvTagLetter } from '@kelvininc/react-ui-components/client';

export const InfoLabelExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvInfoLabel labelTitle="Default" />

		{/*-- Read more and read less --*/}
		<KvInfoLabel
			labelTitle="Description"
			description="Lorem Ipsulum..."
			descriptionHeight="34"
			descriptionCollapsedText="Read more"
			descriptionOpennedText="Read less"
		/>

		{/*-- Copy values --*/}
		<KvInfoLabel
			labelTitle="TYPE"
			description="data-model"
			copyValue="data-model"
		/>

		{/*-- With component --*/}
		<KvInfoLabel labelTitle="DESCRIPTION">
			<KvTagLetter label="Test" tagLetter="T" />
		</KvInfoLabel>
	</>
);

```



## Properties

| Property                   | Attribute                    | Description                                     | Type      | Default                              |
| -------------------------- | ---------------------------- | ----------------------------------------------- | --------- | ------------------------------------ |
| `copyValue`                | `copy-value`                 | (optional) Info label copy value                | `string`  | `undefined`                          |
| `description`              | `description`                | (optional) Info label description               | `string`  | `undefined`                          |
| `descriptionCollapsedText` | `description-collapsed-text` | (optional) Info label description collapse text | `string`  | `DEFAULT_DESCRIPTION_COLLAPSED_TEXT` |
| `descriptionHeight`        | `description-height`         | (optional) Info label description height        | `number`  | `undefined`                          |
| `descriptionOpenedText`    | `description-opened-text`    | (optional) Info label description opened text   | `string`  | `DEFAULT_DESCRIPTION_OPENED_TEXT`    |
| `labelTitle`               | `label-title`                | (optional) Info label title                     | `string`  | `undefined`                          |
| `showTextShadow`           | `show-text-shadow`           | (optional) Show text with a shadow              | `boolean` | `false`                              |


## Shadow Parts

| Part      | Description  |
| --------- | ------------ |
| `"title"` | Label title. |


## CSS Custom Properties

| Name                        | Description                     |
| --------------------------- | ------------------------------- |
| `--description-fade-height` | Description fade height.        |
| `--description-fade-height` | Description fade height.        |
| `--description-fade-shadow` | Description fade shadow.        |
| `--description-fade-shadow` | Description fade shadow.        |
| `--expanded-buttom-color`   | Expandded buttom's color.       |
| `--expanded-buttom-color`   | Expandded buttom's color.       |
| `--text-color-description`  | Info label's description color. |
| `--text-color-description`  | Info label's description color. |
| `--text-color-title`        | Info label's title color.       |
| `--text-color-title`        | Info label's title color.       |


## Dependencies

### Depends on

- [kv-tooltip](../tooltip)
- [kv-icon](../icon)

### Graph
```mermaid
graph TD;
  kv-info-label --> kv-tooltip
  kv-info-label --> kv-icon
  kv-tooltip --> kv-portal
  kv-tooltip --> kv-tooltip-text
  style kv-info-label fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


