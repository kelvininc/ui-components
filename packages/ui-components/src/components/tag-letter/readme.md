# *<kv-tag-letter>*



<!-- Auto Generated Below -->


## Usage

### React

```tsx
import React from 'react';

import { KvTagLetter } from '@kelvininc/react-ui-components/client';

export const TagLetterExample: React.FC = () => (
  <>
    {/*-- Default --*/}
	<KvTagLetter label="Tag Letter" />
	
	{/*-- With Color --*/}
	<KvTagLetter label="Tag Letter" color="green" />

	{/*-- With Tag Letter --*/}
	<KvTagLetter label="Tag Letter" tagLetter="T" />
  </>
);

```



## Properties

| Property             | Attribute    | Description                    | Type     | Default     |
| -------------------- | ------------ | ------------------------------ | -------- | ----------- |
| `color`              | `color`      | (optional) Tag letter's color  | `string` | `undefined` |
| `label` _(required)_ | `label`      | (required) Tag letter's label  | `string` | `undefined` |
| `tagLetter`          | `tag-letter` | (optional) Tag letter's letter | `string` | `undefined` |


----------------------------------------------


