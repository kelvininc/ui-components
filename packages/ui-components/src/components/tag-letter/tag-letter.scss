@import '../../assets/styles/globals';

.tag-letter-container {
	height: 22px;
	border: solid 1px kv-color('primary', 'base');
	border-radius: 2px;
	display: inline-flex;
}

.letter-avatar {
	@include kv-font-h4-semibold();
	height: 100%;
	width: 24px;
	display: flex;
	justify-content: center;

	background-color: kv-color('primary');
	color: kv-color('neutral-0');
}

.label {
	@include kv-font-label-small-regular();
	padding: 0 $spacing;
	align-self: center;
	color: kv-color('text');
}

.tag-letter-container {
	background-color: kv-color('neutral-0');
}
