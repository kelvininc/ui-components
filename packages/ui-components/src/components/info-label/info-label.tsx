import { Component, Prop, h, Host, Element, State } from '@stencil/core';
import { ResizeSensor } from 'css-element-queries';
import { isEmpty } from 'lodash-es';
import { copyTextToClipboard } from '../../utils/clipboard.helper';
import { EIconName } from '../icon/icon.types';
import { DEFAULT_DESCRIPTION_COLLAPSED_TEXT, DEFAULT_DESCRIPTION_OPENED_TEXT } from './info-label.config';
import { COPY_TOOLTIP } from './info-label.types';

/**
 * @part title - Label title.
 */
@Component({
	tag: 'kv-info-label',
	styleUrls: {
		night: 'info-label.night.scss',
		light: 'info-label.light.scss'
	},
	shadow: true
})
export class KvInfoLabel {
	private descriptionContainer: Element;
	private tooltipConfig = COPY_TOOLTIP;
	private resizeSensor: ResizeSensor = null;

	/** Define if the show more button needed to be visible */
	@State() enableShowMoreButton = false;
	/** Define if the label content is completely visible */
	@State() isExpanded = false;
	/** Store the current div content height */
	@State() currentDescriptionHeight: number = null;

	/** (optional) Info label title */
	@Prop({ reflect: true }) labelTitle?: string;
	/** (optional) Info label description */
	@Prop({ reflect: true }) description?: string;
	/** (optional) Info label description height */
	@Prop({ reflect: true }) descriptionHeight?: number;
	/** (optional) Info label description collapse text */
	@Prop({ reflect: true }) descriptionCollapsedText = DEFAULT_DESCRIPTION_COLLAPSED_TEXT;
	/** (optional) Info label description opened text */
	@Prop({ reflect: true }) descriptionOpenedText = DEFAULT_DESCRIPTION_OPENED_TEXT;
	/** (optional) Info label copy value */
	@Prop({ reflect: true }) copyValue?: string;
	/** (optional) Show text with a shadow  */
	@Prop({ reflect: true }) showTextShadow = false;

	/** The Host's element reference */
	@Element() el: HTMLKvInfoLabelElement;

	get showMoreButtonLabel() {
		return this.isExpanded ? this.descriptionOpenedText : this.descriptionCollapsedText;
	}

	private loadDescriptionHeight() {
		if (this.descriptionHeight !== undefined) {
			const descriptionHeight = this.descriptionContainer.clientHeight;
			this.enableShowMoreButton = descriptionHeight !== undefined && descriptionHeight > this.descriptionHeight;
		} else {
			this.enableShowMoreButton = false;
		}
		this.isExpanded = this.enableShowMoreButton ? this.isExpanded : false;
		this.currentDescriptionHeight = this.enableShowMoreButton && !this.isExpanded ? this.descriptionHeight : null;
	}

	private onShowMoreToggle = () => {
		this.isExpanded = !this.isExpanded;

		if (this.isExpanded) {
			const textElementHeight = this.descriptionContainer.clientHeight;
			this.currentDescriptionHeight = textElementHeight;
			return;
		}

		this.currentDescriptionHeight = this.descriptionHeight;
	};

	private onClickCopyAction = async () => {
		const tooltip: HTMLKvTooltipElement = this.el.shadowRoot.querySelector('kv-tooltip');
		const tooltipText = tooltip.shadowRoot.querySelector('#tooltip') as HTMLElement;
		if (await copyTextToClipboard(this.copyValue)) {
			tooltipText.innerText = this.tooltipConfig.resultLabel;
		} else {
			console.error('Copy to clipboard failed');
		}
	};

	componentDidRender() {
		this.descriptionContainer = this.el.shadowRoot.querySelector('.description');

		if (this.descriptionHeight >= 0) {
			this.resizeSensor = new ResizeSensor(this.descriptionContainer, () => {
				this.loadDescriptionHeight();
			});
		}
	}

	disconnectedCallback() {
		this.resizeSensor?.reset();
		this.resizeSensor?.detach();
		this.resizeSensor = null;
	}

	render() {
		return (
			<Host>
				<div class="info-label">
					{this.labelTitle && (
						<div part="title" class={{ 'title': true, 'no-description': isEmpty(this.description) }}>
							{this.labelTitle}
						</div>
					)}
					<div style={{ height: `${this.currentDescriptionHeight}px` }} class="description-wrapper">
						<div class="description">
							<slot>
								{this.description && <div class="text">{this.description}</div>}
								{this.copyValue && (
									<kv-tooltip text={this.tooltipConfig.label} position={this.tooltipConfig.position}>
										<kv-icon class="copy-icon" name={EIconName.Copy} onClick={this.onClickCopyAction} />
									</kv-tooltip>
								)}
							</slot>
						</div>
						{this.showTextShadow && this.enableShowMoreButton && (
							<div
								class={{
									'description-fade': true,
									'description-fade--expanded': this.isExpanded
								}}
							/>
						)}
					</div>
					{this.enableShowMoreButton && (
						<div class={{ 'expand-description-button': true, 'expanded': this.isExpanded }} onClick={this.onShowMoreToggle}>
							{this.showMoreButtonLabel}
							<kv-icon name={EIconName.Expand} />
						</div>
					)}
				</div>
			</Host>
		);
	}
}
