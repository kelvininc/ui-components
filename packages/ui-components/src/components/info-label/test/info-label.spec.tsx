import { SpecPage } from '@stencil/core/internal';
import { KvInfoLabel } from '../info-label';
import { newSpecPage } from '@stencil/core/testing';
import { KvTagLetter } from '../../tag-letter/tag-letter';

describe('Info Label (unit tests)', () => {
	let page: SpecPage;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvInfoLabel],
				html: `<kv-info-label label-text="Default" />`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when renders with description and read more stuffs', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvInfoLabel],
				html: `<kv-info-label
							label-title="DESCRIPTION"
							description="Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book."
							description-height="34"
							description-collapsed-text="Read more"
							description-opened-text="Read less"
							show-text-shadow="true"
						/>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when renders with copy values props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvInfoLabel],
				html: '<kv-info-label label-title="TYPE" description="data-model" copy-value="data-model" />'
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when renders with a component', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvInfoLabel, KvTagLetter],
				html: '<kv-info-label label-title="DESCRIPTION"><kv-tag-letter label="Test" tag-letter="T" /></kv-info-label>'
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
