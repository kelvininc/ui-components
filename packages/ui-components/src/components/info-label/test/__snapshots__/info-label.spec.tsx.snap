// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Info Label (unit tests) when renders with a component should match the snapshot 1`] = `
<kv-info-label description-collapsed-text="Read more" description-opened-text="Read less" label-title="DESCRIPTION">
  <template shadowrootmode="open">
    <div class="info-label">
      <div class="no-description title" part="title">
        DESCRIPTION
      </div>
    </div>
  </template>
  <kv-tag-letter label="Test" tag-letter="T">
    <template shadowrootmode="open">
      <div class="tag-letter-container">
        <div class="letter-avatar">
          <div class="avatar">
            T
          </div>
        </div>
        <div class="label">
          Test
        </div>
      </div>
    </template>
  </kv-tag-letter>
</kv-info-label>
`;

exports[`Info Label (unit tests) when renders with copy values props should match the snapshot 1`] = `
<kv-info-label copy-value="data-model" description="data-model" description-collapsed-text="Read more" description-opened-text="Read less" label-title="TYPE">
  <template shadowrootmode="open">
    <div class="info-label">
      <div class="title" part="title">
        TYPE
      </div>
      <div class="description-wrapper" style="height: nullpx;">
        <div class="description">
          <div class="text">
            data-model
          </div>
          <kv-tooltip position="top" text="Copy">
            <kv-icon class="copy-icon" name="kv-copy"></kv-icon>
          </kv-tooltip>
          <slot></slot>
        </div>
      </div>
    </div>
  </template>
</kv-info-label>
`;

exports[`Info Label (unit tests) when renders with description and read more stuffs should match the snapshot 1`] = `
<kv-info-label description="Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book." description-collapsed-text="Read more" description-height="34" description-opened-text="Read less" label-title="DESCRIPTION">
  <template shadowrootmode="open">
    <div class="info-label">
      <div class="title" part="title">
        DESCRIPTION
      </div>
      <div class="description-wrapper" style="height: nullpx;">
        <div class="description">
          <div class="text">
            Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.
          </div>
          <slot></slot>
        </div>
      </div>
    </div>
  </template>
</kv-info-label>
`;

exports[`Info Label (unit tests) when uses default props should match the snapshot 1`] = `
<kv-info-label description-collapsed-text="Read more" description-opened-text="Read less" label-text="Default">
  <template shadowrootmode="open">
    <div class="info-label"></div>
  </template>
</kv-info-label>
`;
